src/libraries/Microsoft.PowerFx.Interpreter/Functions/LibraryTable.cs

```diff
@@ -294,37 +291,62 @@ private static bool IsValueTypeErrorOrBlank<T>(FormulaValue val)
             return new InMemoryTableValue(irContext, pairs.Select(pair => pair.Key));
         }
 
-        private static async IAsyncEnumerable<DValue<RecordValue>> LazyFilterAsync(
+        private static async Task<DValue<RecordValue>> LazyFilterRowAsync(
+           EvalVisitor runner,
+           SymbolContext context,
+           DValue<RecordValue> row,
+           LambdaFormulaValue filter)
+        {
+            if (row.IsValue)
+            {
+                var childContext = context.WithScopeValues(row.Value);
+
+                // Filter evals to a boolean 
+                var result = await filter.EvalAsync(runner, childContext);
+                var include = false;
+                if (result is BooleanValue booleanValue)
+                {
+                    include = booleanValue.Value;
+                }
+                else if (result is ErrorValue errorValue)
+                {
+                    return DValue<RecordValue>.Of(errorValue);
+                }
+
+                if (include)
+                {
+                    return row;
+                }
+            }
+
+            return null;
+        }
+
+        private static async Task<DValue<RecordValue>[]> LazyFilterAsync(
             EvalVisitor runner,
             SymbolContext context,
             IEnumerable<DValue<RecordValue>> sources,
-            LambdaFormulaValue filter)
+            LambdaFormulaValue filter, 
+            int topN = int.MaxValue)
         {
+            var tasks = new List<Task<DValue<RecordValue>>>();
+            
+            // Filter needs to allow running in parallel. 
             foreach (var row in sources)
             {
                 runner.CheckCancel();
-                if (row.IsValue)
-                {
-                    var childContext = context.WithScopeValues(row.Value);
-
-                    // Filter evals to a boolean 
-                    var result = await filter.EvalAsync(runner, childContext);
-                    var include = false;
-                    if (result is BooleanValue booleanValue)
-                    {
-                        include = booleanValue.Value;
-                    }
-                    else if (result is ErrorValue errorValue)
-                    {
-                        yield return DValue<RecordValue>.Of(errorValue);
-                    }
 
-                    if (include)
-                    {
-                        yield return row;
-                    }
-                }
+                var task = LazyFilterRowAsync(runner, context, row, filter);
+                tasks.Add(task);
             }
+
+            // WhenAll will allow running tasks in parallel. 
+            var results = await Task.WhenAll(tasks);
+
+            // Remove all nulls. 
+            var final = results.Where(x => x != null);
```

Comment:
>var final = results.Where(x => x != null); [](http://example.com/codeflow?start=12&length=42)

This should not be the condition - an expression such as `Filter(Table({a:1},{a:2},If(1<0,{a:3}),{a:4},{a:5}), true)` should return a table with 5 rows (one of them empty).

Reply:
Just noticed that it is a current bug, not related to your changes... Created issue #263 to track it

Reply:
thanks - I believe this is preserving the existing logic (right or wrong) and just making it async.